docs/openstack-cloud-controller-manager/expose-applications-using-loadbalancer-type-service.md

```diff
@@ -180,8 +180,6 @@ Request Body:
 
   Defines whether to create health monitor for the load balancer pool, if not specified, use `create-monitor` config. The health monitor can be created or deleted dynamically. A health monitor is required for services with `externalTrafficPolicy: Local`.
 
-  Not supported when `lb-provider=ovn` is configured in openstack-cloud-controller-manager.
```

Comment:
if the `externalTrafficPolicy: Local` is set, HTTP monitors are being used. If OVN doesn't support HTTP monitors, why do you remove this string?

Reply:
Because the current behaviour is that the health monitor type will initially be set to `port.Protocol`. Later a check will be made if `HTTP` health monitors are supported and if that is the case the type will be changed to `HTTP`, but since OVN does not support it the type will stay at `TCP`.
https://github.com/kubernetes/cloud-provider-openstack/blob/2a74c6c55efba22eec2ad4657025476e5a731347/pkg/openstack/loadbalancer.go#L1153
https://github.com/kubernetes/cloud-provider-openstack/blob/2a74c6c55efba22eec2ad4657025476e5a731347/pkg/openstack/loadbalancer.go#L1162-L1167
Thus the health monitors will be created with TCP which OVN does support (as described [here](https://github.com/kubernetes/cloud-provider-openstack/pull/2234#discussion_r1406114979)). Though there has been an issue with floating IPs but that [issue](https://bugs.launchpad.net/neutron/+bug/1997418) has been fixed earlier this year.

Reply:
Right, so the monitor type will be TCP, which means that `externalTrafficPolicy: Local` is not supported in OVN. The idea of the `externalTrafficPolicy: Local` is to monitor each node that contains a pod's endpoint. Simple TCP check will mark all endpoints as active even though the node doesn't contain a pod.

Reply:
Were you able to capture a healthmonitor check tcpdump on the node host net level?
This is how curl check looks like:

```
curl -s 10.0.13.67:31823 | jq -r
{
  "service": {
    "namespace": "default",
    "name": "nginx"
  },
  "localEndpoints": 1,
  "serviceProxyHealthy": true
}
```

with a 200 response code.

this is how failed check looks like:

```
curl 10.0.13.69:31823
{
        "service": {
                "namespace": "default",
                "name": "nginx"
        },
        "localEndpoints": 0,
        "serviceProxyHealthy": true
}
```

with a 503 response code.

Regular netcat behaves identically for both failed or not failed node:

```
root@busybox:/# echo "" | nc 10.0.13.69 31823
HTTP/1.1 400 Bad Request
Content-Type: text/plain; charset=utf-8
Connection: close

root@busybox:/# echo $?
0
```

Reply:
But what you are connecting to is not the node port but the healthcheck port. This port should always be open. The node port on the other side does behave different. In our case I tested it with an OKD 4.14 cluster with ovn-kubernetes. With `netstat` you can see that the node port is open on every node, but it will reject traffic if no pod is scheduled on that node. I verified that with `nc` days ago and that is also what you can see in the pictures. Only the three nodes running a Nginx ingress controller pod are online. All other nodes are in error state, thus the healthcheck works. And yes `master-2` is in state online for port 80 (node port 30507), but I used `nc` in a loop to connect to the floating ip of the load balancer on port 80 multiple hundred times and never got rejected or timed out (and yes the traffic got routed to all pods). Also the `nc` call directly to the `master-2` on port 80 does get rejected, so it is not a problem of kubernetes, openshift, or ovn-kubernetes, it is a problem of OpenStack and I do not know why or how it is in online state but also does not get selected for routing, but that is also not a problem for OCCM.

Reply:
you're right, I got your point. Basically `kube-proxy` doesn't route the traffic to another nodes, `kubelet` listens the  healtchcheck endpoint, but no healthmonitors check it.

Reply:
Yes, `kube-proxy` does not route the traffic since `externalTrafficPolicy: Local` has been set and the `healthCheckNodePort` is not used. The health monitors directly connect to the node port instead of the `healthCheckNodePort`.

Reply:
@kayrus: Yup, that's how it was designed in https://github.com/kubernetes/cloud-provider-openstack/pull/2000 and I can assure it works as we test this in OpenShift CI.